enterprise/backend/src/metabase_enterprise/workspaces/impl.clj

```diff
@@ -300,16 +301,105 @@
+  "Re-analyze dependencies for any workspace transforms that are stale.
+   Marks them as not stale after analysis."
+  [{ws-id :id, isolated-schema :schema :as workspace}]
+  (let [stale-transforms (t2/select :model/WorkspaceTransform
+                                    :workspace_id ws-id
+                                    :analysis_stale true)]
+    (doseq [transform stale-transforms]
+      (let [analysis (ws.deps/analyze-entity :transform transform)]
+        (ws.deps/write-dependencies! ws-id isolated-schema :transform (:ref_id transform) analysis))
+      (t2/update! :model/WorkspaceTransform (:ref_id transform) {:analysis_stale false}))
+    ;; Grant read access to any new external inputs
+    (when (seq stale-transforms)
+      (sync-grant-accesses! workspace))))
+
+(defn- calculate-and-cache-graph!
+  "Calculate the graph, cache it, sync external inputs/outputs, and mark workspace as not stale."
+  [{ws-id :id, isolated-schema :schema :as workspace}]
+  (t2/with-transaction [_conn]
+    ;; First, re-analyze any stale transforms
+    (analyze-stale-transforms! workspace)
+    ;; Calculate the graph
+    (let [graph (calculate-graph! ws-id)]
+      ;; Sync external outputs and inputs
+      (sync-external-outputs! ws-id isolated-schema (:entities graph))
+      (sync-external-inputs! ws-id (:entities graph))
+      ;; Cache the graph and mark workspace as not stale
+      (t2/update! :model/Workspace ws-id
+                  {:graph          (serialize-graph-for-storage graph)
+                   :analysis_stale false})
+      graph)))
+
 (defn get-or-calculate-graph
-  "Return the graph. Also syncs workspace_output_external and workspace_input_external tables."
-  [{ws-id :id, isolated-schema :schema :as _workspace}]
-  (let [graph (ws.dag/path-induced-subgraph ws-id (t2/select-fn-vec (fn [{:keys [ref_id]}] {:entity-type :transform :id ref_id})
-                                                                    [:model/WorkspaceTransform :ref_id]
-                                                                    :workspace_id ws-id))]
-    (sync-external-outputs! ws-id isolated-schema (:entities graph))
-    (sync-external-inputs! ws-id (:entities graph))
-    graph))
+  "Return the dependency graph for a workspace. Uses cached graph if workspace is not stale,
+   otherwise recalculates it.
+   Also syncs workspace_output_external and workspace_input_external tables when recalculating."
+  [{analysis-stale :analysis_stale, cached-graph :graph :as workspace}]
+  (if-not analysis-stale
+    ;; Return cached graph
+    (deserialize-graph-from-storage cached-graph)
+    ;; Recalculate and cache
+    (calculate-and-cache-graph! workspace)))
+
+(defn mark-transform-stale!
+  "Mark a workspace transform and its workspace as needing analysis recalculation."
+  [workspace-id ref-id]
+  (t2/update! :model/WorkspaceTransform ref-id {:analysis_stale true})
+  (t2/update! :model/Workspace workspace-id {:analysis_stale true}))
```

Comment:
Feels a bit gross having extra statements for this in cases where we're already modifying one or both of them.